### PR TITLE
dialects: (mod_arith) create mod_arith dialect and add initial operation

### PR DESCRIPTION
**Pull request by emmau678**
Create a mod_arith dialect with a basic initial operation, based on the Mod Arith dialect in HEIR.
(link to the original dialect: https://github.com/google/heir/tree/main/lib/Dialect/ModArith)

### CODE REVIEW

File: xdsl/dialects/mod_arith.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from xdsl.irdl import IRDLOperation, irdl_op_definition, operand_def, result_def
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    # assemblyFormat = "operands attr-dict `:` functional-type(operands, results)"
+    pass
+
+
+class BinaryOp(ModArithOp):
```
Comment (tobiasgrosser) [inline xdsl/dialects/mod_arith.py:14]:
```suggestion

class BinaryOp(ModArithOp):
```

Comment (tobiasgrosser) [reply xdsl/dialects/mod_arith.py:14]: Not sure we want double-newlines, but I guess others know the coding style of xDSL better.

File: xdsl/dialects/mod_arith.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        return super().verify_()
+
+
+@irdl_op_definition
+class AddOp(BinaryOp):
+    name = "mod_arith.add"
+
+
+ModArith = Dialect(
+    "modarith",
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:38]:
```suggestion
    "mod_arith",
```

File: xdsl/dialects/mod_arith.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from abc import ABC
+
+from xdsl.dialects.arith import signlessIntegerLike
+from xdsl.ir import Attribute, Dialect, SSAValue
+from xdsl.irdl import IRDLOperation, irdl_op_definition, operand_def, result_def
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    # assemblyFormat = "operands attr-dict `:` functional-type(operands, results)"
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:10]: Please remove old comments

File: xdsl/dialects/mod_arith.py (hunk 1 above)
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:14]: Should this also inherit from `ABC`? (I'm not 100% sure how this works in python so I could be wrong

File: xdsl/dialects/mod_arith.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class BinaryOp(ModArithOp):
+    lhs = operand_def(signlessIntegerLike)
+    rhs = operand_def(signlessIntegerLike)
+    output = result_def(signlessIntegerLike)
+    # assemblyFormat ="$lhs $rhs attr-dict `:` type($output)"  # operands directive not in asm format
+    traits = frozenset((Pure(),))
+
+    def __init__(self, lhs: SSAValue, rhs: SSAValue, result_type: Attribute | None):
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:21]: In a lot of other dialects the type `SSAValue | Operation` is used so that you can pass in single output operations

File: xdsl/dialects/mod_arith.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        if result_type is None:
+            result_type = SSAValue.get(lhs).type
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:23]: Does `lhs.type` not work?

File: xdsl/dialects/mod_arith.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+class ModArithOp(IRDLOperation, ABC):
+    # assemblyFormat = "operands attr-dict `:` functional-type(operands, results)"
+    pass
+
+
+class BinaryOp(ModArithOp):
+    lhs = operand_def(signlessIntegerLike)
+    rhs = operand_def(signlessIntegerLike)
+    output = result_def(signlessIntegerLike)
+    # assemblyFormat ="$lhs $rhs attr-dict `:` type($output)"  # operands directive not in asm format
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:18]: What does this comment mean? Does the assembly format not work?

**Review by alexarice — COMMENTED** [review]
Left a few comments but looks like a good start

File: xdsl/dialects/mod_arith.py (hunk 5 above)
Comment (kimxworrall) [reply xdsl/dialects/mod_arith.py:22]: We copied this from a function in arith - I was presuming it forces the generation of a new SSAValue

File: xdsl/dialects/mod_arith.py (hunk 1 above)
Comment (emmau678) [reply xdsl/dialects/mod_arith.py:13]: I'm not fully sure, do I need to include it again when ModArithOp inherits from ABC?

Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:13]: Yes, in general, every class that is "incomplete" (i.e. should not be instantiated itself but only serves to encapsulate common abstractions) should ideally be marked `ABC` to make it a type error when someone tries to instantiate them.

File: xdsl/dialects/mod_arith.py (hunk 6 above)
Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:17]: Hmm, I think the problem is that the magic fairy dust that is the assembly format printer/parser cannot infer all types with this format. If you want `$lhs`, `$rhs` and `$output` to always have the same type you can use a type variable to achieve this. Happy to show you how that's done!

File: xdsl/dialects/mod_arith.py (hunk 5 above)
Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:27]: `SSAValue.get(...)` in needed to convert from `SSAValue | Operation` to `SSAValue`, so this is the correct approach!

**Review by AntonLydike — COMMENTED** [review]
Welcome @emmau678! How exciting to see the project grow, this is a very strong start already!

**Comment by superlopuh** [comment]
Can you please delete the "Thank you for submitting a PR to xDSL!" bit from the PR description? The PR descriptions get added as commit messages when merged, so it's good to craft these to be helpful when looking at this commit in git history in the future.

File: tests/filecheck/dialects/mod_arith/mod_arith.mlir [hunk 7]
```diff
@@ -0,0 +1,10 @@
+// RUN: XDSL_ROUNDTRIP
+
+"builtin.module"() ({
+  %lhsi1, %rhsi1 = "test.op"() : () -> (i1, i1)
+
+  %add_res = "mod_arith.add"(%lhsi1, %rhsi1) : (i1, i1) -> i1
+
+  // CHECK: %add_res = "mod_arith.add"(%lhsi1, %rhsi1) : (i1, i1) -> i1
+
+}) : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/mod_arith/mod_arith.mlir:10]: I'd suggest a few changes:
```suggestion
// RUN: XDSL_ROUNDTRIP

%lhsi1, %rhsi1 = "test.op"() : () -> (i1, i1)

// CHECK: %add_res = "mod_arith.add"(%lhsi1, %rhsi1) : (i1, i1) -> i1
%add_res = "mod_arith.add"(%lhsi1, %rhsi1) : (i1, i1) -> i1
```
The builtin.module is implied, so we can skip that entirely.
The LLVM approach is usually to add check lines before the thing, as opposed to after. We're not consistent with this, but I've transitioned to this format now, and would recommend it also.

**Comment by tobiasgrosser** [comment]
> Can you please delete the "Thank you for submitting a PR to xDSL!" bit from the PR description? The PR descriptions get added as commit messages when merged, so it's good to craft these to be helpful when looking at this commit in git history in the future.

Actually, I believe that everything below the line won't be included. But it probably does not hurt to clean this up.

File: xdsl/dialects/mod_arith.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/mod_arith.py:13]: This is entirely handled by Ruff :)

File: xdsl/dialects/mod_arith.py (hunk 6 above)
Comment (emmau678) [reply xdsl/dialects/mod_arith.py:17]: that would be great, thanks @AntonLydike

Comment (alexarice) [reply xdsl/dialects/mod_arith.py:17]: Bit of an aside but doesn't `$operands` expand to something more like 
```
`(` $lhs `,` $rhs `)`
```
?

**Review by AntonLydike — APPROVED** [review]
:rocket:

File: xdsl/dialects/mod_arith.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+class BinaryModuloOp(BinaryOp[_T], ABC, Generic[_T]):
+    """
+    Adds the modulus attribute
+    """
+
+    modulus = attr_def(AnyIntegerAttr)
+
+
+@irdl_op_definition
+class AddOp(BinaryModuloOp[Annotated[Attribute, signlessIntegerLike]]):
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:64]: Is there a reason for putting the bound here instead of on the superclass as upstream does?

File: xdsl/dialects/mod_arith.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        rhs: SSAValue | Operation,
+        result_type: Attribute | None,
+    ):
+        if result_type is None:
+            result_type = SSAValue.get(lhs).type
+
+        super().__init__(operands=[lhs, rhs], result_types=[result_type])
+
+
+class BinaryModuloOp(BinaryOp[_T], ABC, Generic[_T]):
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:55]: Why make another class here where upstream just has `ModArith_Op` > `ModArith_BinaryOp` > `ModArith_AddOp`

File: xdsl/dialects/mod_arith.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    Adds the modulus attribute
+    """
+
+    modulus = attr_def(AnyIntegerAttr)
```
Comment (alexarice) [inline xdsl/dialects/mod_arith.py:60]: Tablegen "attributes" are actually properties (not sure if you were trying to be compatible.

**Review by alexarice — COMMENTED** [review]
Some small comments, probably won't block merging

File: xdsl/dialects/mod_arith.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@irdl_op_definition
+class AddOp(BinaryModuloOp[Annotated[Attribute, signlessIntegerLike]]):
+    name = "mod_arith.add"
+
+
+ModArith = Dialect(
+    "mod_arith",
+    [AddOp],
```
Comment (superlopuh) [inline xdsl/dialects/mod_arith.py:70]: Presumably, there will be more ops
```suggestion
    [AddOp,],
```

File: xdsl/dialects/mod_arith.py (hunk 10 above)
Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:60]: Yeah, upstream has modulus as an attribute not a property

File: xdsl/dialects/mod_arith.py (hunk 9 above)
Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:55]: Hmm, not 100% sure what you mean, but there are `ModArith` ops that don't have a modulus.

I was thinking we could eventually model `HasModulus` as a mixin instead of a subclass. I guess this becomes relevant once we get the unary and ternary operators in?

File: xdsl/dialects/mod_arith.py (hunk 8 above)
Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:64]: No particular reason, I guess this becomes visible as duplicated code once we add more ops to the dialect.

File: xdsl/dialects/mod_arith.py (hunk 9 above)
Comment (alexarice) [reply xdsl/dialects/mod_arith.py:55]: In the file I'm looking at, all `ModArith_BinaryOp` have a modulus (maybe I'm looking at an out of date file

File: xdsl/dialects/mod_arith.py (hunk 10 above)
Comment (alexarice) [reply xdsl/dialects/mod_arith.py:60]: In [here](https://github.com/google/heir/blob/main/lib/Dialect/ModArith/IR/ModArithOps.td) I am 90% sure it is a property

Comment (AntonLydike) [reply xdsl/dialects/mod_arith.py:60]: Seeing the assembly format only mentions `attr-dict`:

> `let assemblyFormat = "operands attr-dict `:` type($output)";`

[ModArithOps.td:75](https://github.com/google/heir/blob/4ca0a98a682a0f2f2569bca50007a3b2d5e50ca6/lib/Dialect/ModArith/IR/ModArithOps.td#L75C1-L75C63)

And the op is printed as `%reduce = mod_arith.reduce %c4 { modulus = 17 } : i10` in [tests/mod_arith/syntax.mlir:37](https://github.com/google/heir/blob/4ca0a98a682a0f2f2569bca50007a3b2d5e50ca6/tests/mod_arith/syntax.mlir#L37)

I'm 99.9% sure it's an attribute.

Comment (alexarice) [reply xdsl/dialects/mod_arith.py:60]: I tried to install the tool for an hour and failed, but the real test will be printing generic syntax

Comment (alexarice) [reply xdsl/dialects/mod_arith.py:60]: It took me all evening and installing a virtual machine but can confirm that it is a property:
image

Comment (emmau678) [reply xdsl/dialects/mod_arith.py:60]: I'm not fully clear on what I need to do here - should I change attr_def to prop_def?

File: docs/marimo/linalg_snitch.py
Comment (superlopuh) [inline docs/marimo/linalg_snitch.py:1]: please revert this change

File: docs/marimo/mlir/onnx_demo.py
Comment (superlopuh) [inline docs/marimo/mlir/onnx_demo.py:1]: same here

File: pyproject.toml
Comment (superlopuh) [inline pyproject.toml:1]: this seems unrelated

**Comment by superlopuh** [comment]
Something's off, did a merge go wrong?

**Comment by alexarice** [comment]
> Something's off, did a merge go wrong?

image
Kind of looks like you did this

**Comment by emmau678** [comment]
@superlopuh @alexarice yes, I ran git pull when I should not have, I'm working on it :)

**Comment by emmau678** [comment]
> @superlopuh @alexarice yes, I ran git pull when I should not have, I'm working on it :)

I think I have fixed it now

**Comment by alexarice** [comment]
The merge looks fixed. Just the one comment still up then should be good to merge from my end

**Review by alexarice — APPROVED** [review]

File: xdsl/dialects/mod_arith.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class BinaryOp(ModArithOp, ABC, Generic[_T]):
+    """
+    Simple binary operation
+    """
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
```
Comment (superlopuh) [inline xdsl/dialects/mod_arith.py:36]: Is there any future where the type isn't `signlessIntegerLike`? I'm doing some related work at the moment, and it would make it easier if we could specialize as part of this PR, something like this:
```suggestion
class BinaryOp(ModArithOp, ABC):
    """
    Simple binary operation
    """

    T = Annotated[Attribute, ConstraintVar("T"), signlessIntegerLike]
```

File: xdsl/dialects/mod_arith.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+        super().__init__(
+            operands=[lhs, rhs],
+            result_types=[result_type],
+            properties={"modulus": modulus},
+        )
+
+
+@irdl_op_definition
+class AddOp(BinaryOp[Annotated[Attribute, signlessIntegerLike]]):
```
Comment (superlopuh) [inline xdsl/dialects/mod_arith.py:65]:
```suggestion
class AddOp(BinaryOp):
```
remaining change if you acept the one above

File: xdsl/dialects/mod_arith.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    irdl_op_definition,
+    operand_def,
+    prop_def,
+    result_def,
+)
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    pass
```
Comment (superlopuh) [inline xdsl/dialects/mod_arith.py:25]: I'm not sure that this is necessary, I'd advise against abstracting things before there's a real use for it.

**Review by superlopuh — APPROVED** [review]
